src/sources.rs

```diff
@@ -0,0 +1,346 @@
+use crate::{
+    config,
+    config::{
+        BITFINEX_URL, COINGECKO_URL, COINMARKETCAP_URL, INITIAL_RETRY_INTERVAL, LIVECOINWATCH_URL,
+        MAX_RETRIES,
+    },
+    prometheus,
+};
+use num_rational::BigRational;
+use reqwest::Url;
+use serde::Deserialize as SerdeDeserialize;
+use serde_json::json;
+use std::{
+    collections::VecDeque,
+    future::Future,
+    sync::{Arc, Mutex},
+};
+use tokio::time::{interval, sleep, Duration};
+
+#[derive(Clone)]
+pub enum Source {
+    Bitfinex,
+    Test(Url, String),
+    CoinGecko,
+    LiveCoinWatch(String), // param is api key
+    CoinMarketCap(String), // param is api key
+}
+
+/**
+ * Wrapper for a request function, for continous attempts, with exponential
+ * backoff.
+ * on_fail is invoked when the request_fn fails, but only if there is any
+ * retries left.
+ */
+async fn request_with_backoff<'a, Fut: 'a, T>(
+    request_fn: impl Fn() -> Fut,
+    on_fail: impl Fn(u64),
+    initial_delay: u64,
+    max_retries: u64,
+) -> Option<T>
+where
+    Fut: Future<Output = Option<T>>, {
+    let mut timeout = initial_delay;
+    let mut retries = max_retries;
+    loop {
+        if let Some(i) = request_fn().await {
+            return Some(i);
+        }
+
+        if retries == 0 {
+            return None;
+        }
+
+        on_fail(timeout);
+
+        retries -= 1;
+        sleep(Duration::from_secs(timeout)).await;
+        timeout *= 2;
+    }
+}
+
+/**
+ * Auxillery function for requesting exchange rate.
+ * Handles common behaviour among functions for requesting exchange rate.
+ * The parser should handle converting the JSON response body into an
+ * exchange rate, and its parameter specifies the expected JSON format.
+ */
+async fn request_exchange_rate_aux<
+    ResponseFormat: for<'de> crypto_common::SerdeDeserialize<'de>,
```

Comment:
Since you now depend on Serde, what is the reason for importing serde from crypto_common?

Moreover, you should not import crypto_common directly, it should be imported via rust-sdk.

Reply:
This has been refactored away.
I've fixed the remaining imports.